components/merkledb/src/proof_list_index/mod.rs

```diff
@@ -548,21 +657,77 @@ where
                 index
             );
         }
-        let mut key = ProofListKey::new(1, index);
-        self.base.put(&key, HashTag::hash_leaf(&value.to_bytes()));
+        self.base.put(
+            &ProofListKey::new(1, index),
+            HashTag::hash_leaf(&value.to_bytes()),
+        );
         self.base.put(&ProofListKey::leaf(index), value);
-        while key.height() < self.height() {
-            let (left, right) = (key.as_left(), key.as_right());
-            let hash = if self.has_branch(right) {
-                HashTag::hash_node(
-                    &self.get_branch_unchecked(left),
-                    &self.get_branch_unchecked(right),
-                )
-            } else {
-                HashTag::hash_single_node(&self.get_branch_unchecked(left))
-            };
-            key = key.parent();
-            self.set_branch(key, hash);
+        self.update_range(index, index);
+    }
+
+    /// Shortens the list, keeping the indicated number of first `len` elements
+    /// and dropping the rest.
+    ///
+    /// If `len` is greater than the current state of the list, this has no effect.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum_merkledb::{TemporaryDB, Database, ProofListIndex};
+    ///
+    /// let db = TemporaryDB::new();
+    /// let name = "name";
+    /// let fork = db.fork();
+    /// let mut index = ProofListIndex::new(name, &fork);
+    ///
+    /// index.extend([1, 2, 3, 4, 5].iter().cloned());
+    /// assert_eq!(5, index.len());
+    /// index.truncate(3);
+    /// assert!(index.iter().eq(vec![1, 2, 3]));
+    /// ```
+    pub fn truncate(&mut self, new_length: u64) {
+        if self.len() <= new_length {
+            return;
+        }
+        if new_length == 0 {
+            self.clear();
+            return;
+        }
+
+        let old_last_index = self.len() - 1;
+        let old_height = self.height();
+        self.set_len(new_length);
+
+        // Remove values.
+        for index in new_length..=old_last_index {
+            self.base.remove(&ProofListKey::leaf(index));
+        }
+
+        self.remove_range(old_last_index, old_height);
+    }
+
+    /// Removes the last element from the list and returns it, or returns `None`
```

Comment:
I suppose that data structures, such as the proof list index, are often used as append only structures.
So that do you think about creating append only wrapper for this structure with disallowed mutation operations.

Reply:
Why? If service author don't want to remove anything from the index, he just won't call those methods.

We should not protect users like "oh I accidently called `index.clear()`", we're not babysitters.

Reply:
IMO I don't see a reason why to do so. It's a common method for DS with access to its elements by index. So it's ok as for me.